lib/client.go

```diff
@@ -27,28 +30,127 @@ type Client struct {
 	// ServerIp is the public IPv4 address of the server.
 	ServerIp netip.Addr
 
-	// Password is authenticates the client connection.
+	// Password authenticates the client connection.
 	Password string
 
 	// WgClient is a shared client for interacting with the WireGuard kernel module.
 	WgClient *wgctrl.Client
 
 	// Http is used to make connect requests to the server.
 	Http *http.Client
+
+	// wgCidr is the current subnet assigned to the WireGuard interface, if any.
+	wgCidr netip.Prefix
+}
+
+// Connect attempts to set up the connection with the peer and create a network
+// interface for it. Disconnect() needs to be called when the connection is no
+// longer needed to clean up the created network interface.
+func (c *Client) Connect() error {
+	connectionResponse, err := c.sendConnectionRequest()
+	if err != nil {
+		return err
+	}
+
+	err = c.createInterface(connectionResponse)
+	if err != nil {
+		return err
+	}
+
+	err = c.configureWireguard(connectionResponse)
+	if err != nil {
+		netlink.LinkDel(c.link())
+		return fmt.Errorf("error configuring vprox interface: %v", err)
+	}
+
+	return nil
+}
+
+// Connect attempts to reconnect to the peer, reusing the existing network
+// interface.
+func (c *Client) Reconnect() error {
+	resp, err := c.sendConnectionRequest()
+	if err != nil {
+		return err
+	}
+
+	err = c.updateInterface(resp)
+	if err != nil {
+		return err
+	}
+
+	err = c.configureWireguard(resp)
+	if err != nil {
+		return fmt.Errorf("error reconfiguring vprox interface: %v", err)
+	}
+
+	return nil
+}
+
+// createInterface creates a new WireGuard interface.
+func (c *Client) createInterface(connectionResponse connectResponse) error {
+	link := c.link()
+
+	err := netlink.LinkAdd(link)
+	if err != nil {
+		return fmt.Errorf("error creating vprox interface: %v", err)
+	}
+
+	err = c.updateInterface(connectionResponse)
+	if err != nil {
+		netlink.LinkDel(link)
+		return err
+	}
+
+	err = netlink.LinkSetUp(link)
+	if err != nil {
+		netlink.LinkDel(link)
+		return fmt.Errorf("error setting up vprox interface: %v", err)
+	}
+	return nil
 }
 
-func (c *Client) CreateInterface() error {
+// updateInterface updates the wireguard interface based on the provided connectionResponse
+func (c *Client) updateInterface(resp connectResponse) error {
+	cidr, err := netip.ParsePrefix(resp.AssignedAddr)
+	if err != nil {
+		return fmt.Errorf("failed to parse assigned address %v: %v", resp.AssignedAddr, err)
+	}
+
+	if cidr != c.wgCidr {
+		link := c.link()
+
+		if c.wgCidr != (netip.Prefix{}) {
+			oldIpnet := prefixToIPNet(c.wgCidr)
+			err = netlink.AddrDel(link, &netlink.Addr{IPNet: &oldIpnet})
+			if err != nil {
+				return fmt.Errorf("failed to remove old address from vprox interface: %v", err)
+			}
```

Comment:
If err is that the address does not exist, we should continue gracefully here as that may indicate that the state is out of sync, rather than trying to reconnect indefinitely. It looks like the [error is created here](https://github.com/vishvananda/netlink/blob/084abd93d350e97ee5410b5b6311bcc211f7ea05/nl/nl_linux.go#L626) and you can detect if it is the EEXIST.